packages/typescript-estree/src/check-syntax-errors.ts

```diff
@@ -13,14 +18,351 @@ export function checkSyntaxError(tsNode: ts.Node): void {
+          "Classes may not have a field named 'constructor'.",
+        );
+      }
+      break;
+    }
+
+    case SyntaxKind.TaggedTemplateExpression:
+      if (node.tag.flags & ts.NodeFlags.OptionalChain) {
+        throw createError(
+          node,
+          'Tagged template expressions are not permitted in an optional chain.',
+        );
+      }
+      break;
+
+    case SyntaxKind.ClassDeclaration:
+      if (
+        !node.name &&
+        (!hasModifier(ts.SyntaxKind.ExportKeyword, node) ||
+          !hasModifier(ts.SyntaxKind.DefaultKeyword, node))
+      ) {
+        throw createError(
+          node,
+          "A class declaration without the 'default' modifier must have a name.",
+        );
+      }
+      break;
+
+    case SyntaxKind.BinaryExpression:
+      if (
+        node.operatorToken.kind !== SyntaxKind.InKeyword &&
+        node.left.kind === SyntaxKind.PrivateIdentifier
+      ) {
+        throw createError(
+          node.left,
+          "Private identifiers cannot appear on the right-hand-side of an 'in' expression.",
+        );
+      } else if (node.right.kind === SyntaxKind.PrivateIdentifier) {
+        throw createError(
+          node.right,
+          "Private identifiers are only allowed on the left-hand-side of an 'in' expression.",
+        );
+      }
+      break;
+
+    case SyntaxKind.MappedType:
+      if (node.members && node.members.length > 0) {
+        throw createError(
+          node.members[0],
+          'A mapped type may not declare properties or methods.',
+        );
+      }
+      break;
+
+    case SyntaxKind.PropertySignature:
+      // eslint-disable-next-line @typescript-eslint/no-deprecated
+      const { initializer } = node;
+      if (initializer) {
+        throw createError(
+          initializer,
+          'A property signature cannot have an initializer.',
+        );
+      }
+      break;
+
+    case SyntaxKind.FunctionType: {
+      // eslint-disable-next-line @typescript-eslint/no-deprecated
+      const { modifiers } = node;
+      if (modifiers) {
+        throw createError(
+          modifiers[0],
+          'A function type cannot have modifiers.',
+        );
+      }
+      break;
+    }
+
+    case SyntaxKind.EnumMember: {
+      const computed = node.name.kind === ts.SyntaxKind.ComputedPropertyName;
+      if (computed) {
+        throw createError(
+          node.name,
+          'Computed property names are not allowed in enums.',
+        );
+      }
+
+      if (
+        node.name.kind === SyntaxKind.NumericLiteral ||
+        node.name.kind === SyntaxKind.BigIntLiteral
+      ) {
+        throw createError(
+          node.name,
+          'An enum member cannot have a numeric name.',
+        );
+      }
+      break;
+    }
+
+    case SyntaxKind.ExternalModuleReference:
+      if (node.expression.kind !== SyntaxKind.StringLiteral) {
+        throw createError(node.expression, 'String literal expected.');
+      }
+      break;
```

Comment:
@JoshuaKGoldberg Now you can see why I insist one single call in convert.ts, otherwise we need to create and call so many functions.

Reply:
Ha, yeah, I do see what you mean!